app/uk/gov/hmrc/economiccrimelevyreturns/services/SessionService.scala

```diff
@@ -16,53 +16,142 @@
-    sessionRetrievalConnector.delete(internalId)
+  )(implicit hc: HeaderCarrier): EitherT[Future, SessionError, Unit] =
+    EitherT {
+      sessionRetrievalConnector
+        .delete(internalId)
+        .map {
+          Right(_)
+        }
+        .recover {
+          case error @ UpstreamErrorResponse(message, code, _, _)
+              if UpstreamErrorResponse.Upstream5xxResponse
+                .unapply(error)
+                .isDefined || UpstreamErrorResponse.Upstream4xxResponse.unapply(error).isDefined =>
+            Left(SessionError.BadGateway(message, code))
+          case NonFatal(thr) =>
+            Left(SessionError.InternalUnexpectedError(thr.getMessage, Some(thr)))
+        }
+    }
+
+  private def getSessionData(
+    internalId: String
+  )(implicit hc: HeaderCarrier): EitherT[Future, SessionError, Option[SessionData]] =
+    EitherT {
+      sessionRetrievalConnector
+        .get(internalId)
+        .map(s => Right(Some(s)))
+        .recover {
+          case _: NotFoundException                         => Right(None)
+          case _ @UpstreamErrorResponse(_, NOT_FOUND, _, _) =>
+            Right(None)
+          case error @ UpstreamErrorResponse(message, code, _, _)
+              if UpstreamErrorResponse.Upstream5xxResponse
+                .unapply(error)
+                .isDefined || UpstreamErrorResponse.Upstream4xxResponse.unapply(error).isDefined =>
+            Left(SessionError.BadGateway(message, code))
+          case NonFatal(thr)                                =>
+            Left(SessionError.InternalUnexpectedError(thr.getMessage, Some(thr)))
+        }
+    }
+
+  private def retrieveValueFromSessionData(
+    sessionData: SessionData,
+    key: String
+  ): EitherT[Future, SessionError, String] =
+    EitherT {
+      sessionData.values.get(key) match {
+        case Some(value) => Future.successful(Right(value))
+        case None        =>
+          Future.successful(Left(SessionError.KeyNotFound(key)))
+      }
+    }
+
+  private def upsertSessions(sessionData: SessionData)(implicit
+    hc: HeaderCarrier
+  ): EitherT[Future, SessionError, Unit] =
+    EitherT {
+      sessionRetrievalConnector
+        .upsert(sessionData)
+        .map(Right(_))
+        .recover {
+          case error @ UpstreamErrorResponse(message, code, _, _)
+              if UpstreamErrorResponse.Upstream5xxResponse.unapply(error).isDefined ||
+                UpstreamErrorResponse.Upstream4xxResponse.unapply(error).isDefined =>
+            Left(SessionError.BadGateway(message, code))
+          case NonFatal(thr) => Left(SessionError.InternalUnexpectedError(thr.getMessage, Some(thr)))
+        }
+    }
 
+  private def mergeSessionData(oldSessionData: Option[SessionData], newSessionData: SessionData): SessionData =
+    oldSessionData
+      .map(oldData =>
+        newSessionData.copy(
+          values = newSessionData.values ++ oldData.values.filter(e => !newSessionData.values.keySet.contains(e._1))
```

Comment:
To prevent using `!` you could use `filterNot` instead: 

`oldData.values.filterNot(e => newSessionData.values.keySet.contains(e._1))`